upstairs/src/lib.rs

```diff
@@ -1506,56 +1507,140 @@ impl Buffer {
         self.len == 0
     }
 
-    pub async fn as_vec(&self) -> MutexGuard<'_, Vec<u8>> {
-        self.data.lock().await
+    pub fn write(&mut self, offset: usize, data: &[u8]) {
+        assert!(offset + data.len() <= self.data.len());
+        for i in 0..data.len() {
+            self.data[offset + i] = data[i];
+            self.owned[offset + i] = true;
+        }
     }
 
-    pub async fn owned_vec(&self) -> MutexGuard<'_, Vec<bool>> {
-        self.owned.lock().await
+    pub fn write_with_ownership(
+        &mut self,
+        offset: usize,
+        data: &[u8],
+        owned: &[bool],
+    ) {
+        assert!(offset + data.len() <= self.data.len());
+        for i in 0..data.len() {
+            if owned[i] {
+                self.data[offset + i] = data[i];
+                self.owned[offset + i] = true;
+            }
+        }
+    }
+
+    pub fn write_read_response(
+        &mut self,
+        offset: usize,
+        response: &ReadResponse,
+    ) {
+        assert!(offset + response.data.len() <= self.data.len());
+        if !response.block_contexts.is_empty() {
+            for i in 0..response.data.len() {
+                self.data[offset + i] = response.data[i];
```

Comment:
Same here (using `copy_from_slice` / `fill` instead of explicit loop)

Reply:
same in 9cd5f29